Subject: [PATCH] refpolicy: fix modular issue for xconsole

xconsole is defined in xserver module, but logging module uses it
for syslogd domains to access /dev/xconsole.

Move it to logging module since logging is a core module but
xserver not.

Upstream-Status: pending

Signed-off-by: Xin Ouyang <Xin.Ouyang@windriver.com>
Signed-off-by: Wenzong Fan <wenzong.fan@windriver.com>
---
 policy/modules/services/xserver.fc |  5 ----
 policy/modules/services/xserver.if | 36 -------------------------
 policy/modules/services/xserver.te |  9 ++-----
 policy/modules/system/init.te      |  4 ++-
 policy/modules/system/logging.fc   |  2 ++
 policy/modules/system/logging.if   | 54 ++++++++++++++++++++++++++++++++++++++
 policy/modules/system/logging.te   | 15 +++++++----
 7 files changed, 71 insertions(+), 54 deletions(-)

--- refpolicy.orig/policy/modules/services/xserver.fc
+++ refpolicy/policy/modules/services/xserver.fc
@@ -13,11 +13,6 @@ HOME_DIR/\.xsession-errors	--	gen_contex
 HOME_DIR/\.Xauthority.*	--	gen_context(system_u:object_r:xauth_home_t,s0)
 
 #
-# /dev
-#
-/dev/xconsole		-p	gen_context(system_u:object_r:xconsole_device_t,s0)
-
-#
 # /etc
 #
 /etc/gdm(3)?/PostSession/.* --	gen_context(system_u:object_r:xsession_exec_t,s0)
--- refpolicy.orig/policy/modules/services/xserver.if
+++ refpolicy/policy/modules/services/xserver.if
@@ -664,42 +664,6 @@ interface(`xserver_read_user_iceauth',`
 
 ########################################
 ## <summary>
-##	Set the attributes of the X windows console named pipes.
-## </summary>
-## <param name="domain">
-##	<summary>
-##	Domain allowed access.
-##	</summary>
-## </param>
-#
-interface(`xserver_setattr_console_pipes',`
-	gen_require(`
-		type xconsole_device_t;
-	')
-
-	allow $1 xconsole_device_t:fifo_file setattr;
-')
-
-########################################
-## <summary>
-##	Read and write the X windows console named pipe.
-## </summary>
-## <param name="domain">
-##	<summary>
-##	Domain allowed access.
-##	</summary>
-## </param>
-#
-interface(`xserver_rw_console',`
-	gen_require(`
-		type xconsole_device_t;
-	')
-
-	allow $1 xconsole_device_t:fifo_file rw_fifo_file_perms;
-')
-
-########################################
-## <summary>
 ##	Use file descriptors for xdm.
 ## </summary>
 ## <param name="domain">
--- refpolicy.orig/policy/modules/services/xserver.te
+++ refpolicy/policy/modules/services/xserver.te
@@ -159,13 +159,6 @@ typealias xauth_tmp_t alias { user_xauth
 typealias xauth_tmp_t alias { auditadm_xauth_tmp_t secadm_xauth_tmp_t };
 userdom_user_tmp_file(xauth_tmp_t)
 
-# this is not actually a device, its a pipe
-type xconsole_device_t;
-files_type(xconsole_device_t)
-dev_associate(xconsole_device_t)
-fs_associate_tmpfs(xconsole_device_t)
-files_associate_tmp(xconsole_device_t)
-
 type xdm_t;
 type xdm_exec_t;
 auth_login_pgm_domain(xdm_t)
@@ -329,6 +322,8 @@ allow xdm_t self:appletalk_socket create
 allow xdm_t self:key { search link write };
 
 allow xdm_t xconsole_device_t:fifo_file { read_fifo_file_perms setattr_fifo_file_perms };
+logging_setattr_xconsole(xdm_t)
+logging_getattr_xconsole(xdm_t)
 
 # Allow gdm to run gdm-binary
 can_exec(xdm_t, xdm_exec_t)
--- refpolicy.orig/policy/modules/system/init.te
+++ refpolicy/policy/modules/system/init.te
@@ -1121,8 +1121,10 @@ optional_policy(`
 
 optional_policy(`
 	# Set device ownerships/modes.
-	xserver_setattr_console_pipes(initrc_t)
+	logging_setattr_xconsole(initrc_t)
+')
 
+optional_policy(`
 	# init script wants to check if it needs to update windowmanagerlist
 	xserver_read_xdm_rw_config(initrc_t)
 ')
--- refpolicy.orig/policy/modules/system/logging.fc
+++ refpolicy/policy/modules/system/logging.fc
@@ -1,5 +1,7 @@
 /dev/log		-s	gen_context(system_u:object_r:devlog_t,mls_systemhigh)
 
+/dev/xconsole           -p      gen_context(system_u:object_r:xconsole_device_t,s0)
+
 /etc/rsyslog.conf		gen_context(system_u:object_r:syslog_conf_t,s0)
 /etc/syslog.conf		gen_context(system_u:object_r:syslog_conf_t,s0)
 /etc/syslog.conf\.sysklogd	gen_context(system_u:object_r:syslog_conf_t,s0)
--- refpolicy.orig/policy/modules/system/logging.if
+++ refpolicy/policy/modules/system/logging.if
@@ -1017,6 +1017,60 @@ interface(`logging_manage_generic_logs',
 
 ########################################
 ## <summary>
+##     Set the attributes of the X windows console named pipes.
+## </summary>
+## <param name="domain">
+##     <summary>
+#      Domain allowed access.
+##     </summary>
+## </param>
+#
+interface(`logging_setattr_xconsole',`
+	gen_require(`
+		type xconsole_device_t;
+	')
+
+	allow $1 xconsole_device_t:fifo_file setattr_fifo_file_perms;
+')
+
+########################################
+## <summary>
+##     Set the attributes of the X windows console named pipes.
+## </summary>
+## <param name="domain">
+##     <summary>
+#      Domain allowed access.
+##     </summary>
+## </param>
+#
+interface(`logging_getattr_xconsole',`
+	gen_require(`
+		type xconsole_device_t;
+	')
+
+	allow $1 xconsole_device_t:fifo_file getattr_fifo_file_perms;
+')
+
+########################################
+## <summary>
+##     Read and write the X windows console named pipe.
+## </summary>
+## <param name="domain">
+##     <summary>
+##     Domain allowed access.
+##     </summary>
+## </param>
+#
+interface(`logging_rw_xconsole',`
+	gen_require(`
+		type xconsole_device_t;
+	')
+
+	allow $1 xconsole_device_t:fifo_file rw_fifo_file_perms;
+')
+
+########################################
+## <summary>
 ##	All of the rules required to administrate
 ##	the audit environment
 ## </summary>
--- refpolicy.orig/policy/modules/system/logging.te
+++ refpolicy/policy/modules/system/logging.te
@@ -89,6 +89,13 @@ type var_log_t;
 logging_log_file(var_log_t)
 files_mountpoint(var_log_t)
 
+# this is not actually a device, its a pipe
+type xconsole_device_t;
+files_type(xconsole_device_t)
+dev_node(xconsole_device_t)
+fs_associate_tmpfs(xconsole_device_t)
+files_associate_tmp(xconsole_device_t)
+
 ifdef(`enable_mls',`
 	init_ranged_daemon_domain(auditd_t, auditd_exec_t, mls_systemhigh)
 	init_ranged_daemon_domain(syslogd_t, syslogd_exec_t, mls_systemhigh)
@@ -614,8 +621,6 @@ optional_policy(`
 	udev_read_db(syslogd_t)
 ')
 
-optional_policy(`
-	# log to the xconsole
-	xserver_rw_console(syslogd_t)
-	xserver_setattr_console_pipes(syslogd_t)
-')
+# log to the xconsole
+logging_rw_xconsole(syslogd_t)
+logging_setattr_xconsole(syslogd_t)
